internal/eventstream/client_api.go

```diff
@@ -0,0 +1,438 @@
+// JSON response - passed as an io.ReaderCloser which will be closed within this function - into the typed response to
+// be used within the methods below.
+//
+//	[0] https://github.com/golang/go/issues/49085
+func extractObjectQueriesResult[T Comment | HostServiceRuntimeAttributes](jsonResp io.ReadCloser) ([]ObjectQueriesResult[T], error) {
+	defer func() {
+		_, _ = io.Copy(io.Discard, jsonResp)
+		_ = jsonResp.Close()
+	}()
+
+	var objQueriesResults []ObjectQueriesResult[T]
+	err := json.NewDecoder(jsonResp).Decode(&struct {
+		Results *[]ObjectQueriesResult[T] `json:"results"`
+	}{&objQueriesResults})
+	if err != nil {
+		return nil, err
+	}
+	return objQueriesResults, nil
+}
+
+// queryObjectsApi performs a configurable HTTP request against the Icinga 2 API and returns its raw response.
+//
+// The returned io.ReaderCloser MUST be both read to completion and closed to reuse connections.
+func (client *Client) queryObjectsApi(
+	ctx context.Context,
+	urlPaths []string,
+	method string,
+	body io.Reader,
+	headers map[string]string,
+) (io.ReadCloser, error) {
+	apiUrl, err := url.JoinPath(client.ApiBaseURL, urlPaths...)
+	if err != nil {
+		return nil, err
+	}
+	req, err := http.NewRequestWithContext(ctx, method, apiUrl, body)
+	if err != nil {
+		return nil, err
+	}
+
+	req.SetBasicAuth(client.ApiBasicAuthUser, client.ApiBasicAuthPass)
+	for k, v := range headers {
+		req.Header.Set(k, v)
+	}
+
+	// The underlying network connection is reused by using client.ApiHttpTransport.
+	httpClient := &http.Client{
+		Transport: &client.ApiHttpTransport,
+		Timeout:   3 * time.Second,
+	}
+	res, err := httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	if res.StatusCode != http.StatusOK {
+		_, _ = io.Copy(io.Discard, res.Body)
+		_ = res.Body.Close()
+		return nil, fmt.Errorf("unexpected HTTP status code %d", res.StatusCode)
+	}
+
+	return res.Body, nil
+}
+
+// queryObjectsApiDirect performs a direct resp. "fast" API query against an object, optionally identified by its name.
+func (client *Client) queryObjectsApiDirect(ctx context.Context, objType, objName string) (io.ReadCloser, error) {
+	return client.queryObjectsApi(
+		ctx,
+		[]string{"/v1/objects/", objType + "s/", rawurlencode(objName)},
+		http.MethodGet,
+		nil,
+		map[string]string{"Accept": "application/json"})
+}
+
+// queryObjectsApiQuery sends a query to the Icinga 2 API /v1/objects to receive data of the given objType.
+func (client *Client) queryObjectsApiQuery(ctx context.Context, objType string, query map[string]any) (io.ReadCloser, error) {
+	reqBody, err := json.Marshal(query)
+	if err != nil {
+		return nil, err
+	}
+
+	return client.queryObjectsApi(
+		ctx,
+		[]string{"/v1/objects/", objType + "s"},
+		http.MethodPost,
+		bytes.NewReader(reqBody),
+		map[string]string{
+			"Accept":                 "application/json",
+			"Content-Type":           "application/json",
+			"X-Http-Method-Override": "GET",
+		})
+}
```

Comment:
This has sent me deep down a rabbit hole…

Yes, in theory or when directly speaking to the Icinga 2 API, this will work - as I have also verified this. However, when some industry grade firewall/proxy/middleware thingy appears, this will might stop working as I have tried to explain in <https://github.com/Icinga/icinga2/pull/9958>.
Thus, I would like to leave the code as it is, as otherwise strange bugs might appear in the future.